### PR TITLE
feat: Codemods for Upgrading from Preact 8.x

### PR DESCRIPTION
**Pull request by manishjha-04**
**🚀 Feature Proposal**

Building codemods for Upgrading from Preact 8.x
Looking forward to add these and help Preact users have a smooth migration experience.

**Motivation**
The motivation to create codemods is that it would streamline the process for millions of users, and also ensure that the application stays efficient and secure while leveraging the latest advancements in Preact.

**Example**
These codemods would help migrate from Preact 8.x  real quick as running :

> `npx codemod preact/X/migration-recipe `

which would include all the codemods.

### CODE REVIEW

File: content/en/guide/v10/upgrade-guide.md [hunk 1]
```diff
@@ -220,6 +255,12 @@ We renamed/moved the following properties:
 
 As much as we tried, we always ran into edge-cases with third-party libraries written for react. This change to our `vnode` shape removed many difficult to spot bugs and makes our `compat` code a lot cleaner.
 
+**Codemod** for renaming properties
+```
+ npx codemod preact/X/libraryAuthors-attribute-node-children-renaming
+```
+
+
```
Comment (mohab-sameh) [inline content/en/guide/v10/upgrade-guide.md:263]: @manishjha-04 please remove this extra line

**Comment by rschristian** [comment]
I'm not sure we need this? Preact X has been out for a lot of years now, providing further migration assistance from 8 at this point doesn't seem necessary.

File: content/en/guide/v10/upgrade-guide.md [hunk 2]
```diff
@@ -78,6 +78,25 @@ The `Portal` component is now part of `preact/compat`.
...
 
 ## Getting your code ready
+---
+### Codemods
+
+To assist with the upgrade, a set of codemods is available to automatically update your codebase from Preact 8.x. These codemods automate common code transformations to streamline the migration process.
+
+Run all codemods listed in this guide with the Preact X codemod recipe:
+
+  ```     npx codemod preact/X/migration-recipe```
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:88]:
```suggestion
```bash
npx codemod preact/X/migration-recipe
```
```

File: content/en/guide/v10/upgrade-guide.md [hunk 3]
```diff
@@ -96,6 +115,11 @@ import { h, Component } from "preact";
 
 _Note: This change doesn't affect `preact/compat`. It still has both named and a default export to remain compatible with react._
 
+**Codemod** for using named exports
+```
+  npx codemod preact/X/default-import-to-namespace-import
+```
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:121]:
```suggestion

```bash
npx codemod preact/X/default-import-to-namespace-import
```
```

File: content/en/guide/v10/upgrade-guide.md [hunk 4]
```diff
@@ -178,6 +202,11 @@ function Foo(props) {
 }
 ```
 
+**Codemod** for `props.children` to `childArray`
+```
+ npx codemod preact/X/props-children-to-child-array
+```
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:208]:
```suggestion

```bash
npx codemod preact/X/props-children-to-child-array
```
```

File: content/en/guide/v10/upgrade-guide.md [hunk 5]
```diff
@@ -195,6 +224,11 @@ this.setState(prevState => {
 });
 ```
 
+**Codemod** for `setState` callback
+```
+ npx codemod preact/X/replace-this-state-with-prevstate
+```
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:230]:
```suggestion

```bash
 npx codemod preact/X/replace-this-state-with-prevstate
```
```

File: content/en/guide/v10/upgrade-guide.md [hunk 6]
```diff
@@ -220,6 +254,10 @@ We renamed/moved the following properties:
 
 As much as we tried, we always ran into edge-cases with third-party libraries written for react. This change to our `vnode` shape removed many difficult to spot bugs and makes our `compat` code a lot cleaner.
 
+**Codemod** for renaming properties
+```
+ npx codemod preact/X/libraryAuthors-attribute-node-children-renaming
+```
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:260]:
```suggestion

```bash
npx codemod preact/X/libraryAuthors-attribute-node-children-renaming
```
```

File: content/en/guide/v10/upgrade-guide.md [hunk 7]
```diff
@@ -78,6 +78,25 @@ The `Portal` component is now part of `preact/compat`.
 2. Import `createPortal` from `preact/compat`
 
 ## Getting your code ready
+---
+### Codemods
```
Comment (rschristian) [inline content/en/guide/v10/upgrade-guide.md:82]:
```suggestion
### Codemods
```

This should probably come last on the page

**Comment by mohab-sameh** [comment]
> Preact X has been out for over 5 years now, with [Preact 8 having pretty few users](https://majors.nullvoxpopuli.com/q?minors=&old=&packages=preact).
> 
> I'm not sure this is really needed, and we'd probably need to do some testing before recommending it in our docs.

@rschristian Thanks for the helpful insights. In this case, we can help make codemods for a migration that is more helpful and in need. If you have suggestions you'd like to see, please let me know.

Also, we should definitely pressure test any codemods before merging them to the upgrade doc. @manishjha-04 can you please turn this into a draft PR until everything is pressure tested?

**Comment by rschristian** [comment]
> In this case, we can help make codemods for a migration that is more helpful and in need. If you have suggestions you'd like to see, please let me know.

Are you actively trying to make something Preact-related? Just curious, as the PR is a bit oddly timed given the age of the target (Preact X). That not necessarily a problem by any means, just wondering if this is a result of actively trying to make a Preact-specific code mod?

A major point of our ecosystem at the moment is [a lack of breaking changes](https://preactjs.com/blog/preact-x), so I don't think there's anything really out there. I suppose there's [`preact-router`](https://github.com/preactjs/preact-router) to [`preact-iso`](https://github.com/preactjs/preact-iso), but it's not 1:1, so I'm not sure you could automate it, nor do I think there's a big audience there ncessarily. That'd be the only thing that comes to mind.

**Comment by mohab-sameh** [comment]
> > In this case, we can help make codemods for a migration that is more helpful and in need. If you have suggestions you'd like to see, please let me know.
> 
> Are you actively trying to make something Preact-related? Just curious, as the PR is a bit oddly timed given the age of the target (Preact X). That not necessarily a problem by any means, just wondering if this is a result of actively trying to make a Preact-specific code mod?
> 
> A major point of our ecosystem at the moment is [a lack of breaking changes](https://preactjs.com/blog/preact-x), so I don't think there's anything really out there. I suppose there's [`preact-router`](https://github.com/preactjs/preact-router) to [`preact-iso`](https://github.com/preactjs/preact-iso), but it's not 1:1, so I'm not sure you could automate it, nor do I think there's a big audience there ncessarily. That'd be the only thing that comes to mind.

We had an interest in our community in codemodding low-hanging fruits that help improve OSS framework users' upgrade experience, and Preact had a clear and codemoddeable migration doc. I think we should've taken into account that such an older version may not be widely adopted anymore. Yeah, it doesn't hurt to have these anyway, but I agree that it makes more sense to tackle something more impactful.

Well, since we're at it, if you come across any best-practice or repetitive refactors you'd like to see, please let us know ([our Slack community](https://go.codemod.com/community)). Meanwhile, we can test these codemods and put them out there. Hopefully they might help someone one day.

**Comment by rschristian** [comment]
I'm going to close this on out. Due to the age of Preact X it's going to be pretty niche if it gets any uses at all and I'm generally uncomfortable pushing users toward arbitrary scripts that we don't control ourselves.

Thanks though.

**Pull request closed without merging**